src/anbox/cmds/launch.cpp

```diff
@@ -69,51 +86,95 @@ anbox::cmds::Launch::Launch()
+
+        std::map<std::string,std::string> env;
+        core::posix::this_process::env::for_each([&](const std::string &name, const std::string &value) {
+          env.insert({name, value});
+        });
+
+        const auto exe_path = utils::process_get_exe_path(::getpid());
+        if (!fs::exists(exe_path)) {
+          ERROR("Can't find correct anbox executable to run. Found %s but does not exist", exe_path);
+          return EXIT_FAILURE;
+        }
+
+        try {
+          const auto flags = core::posix::StandardStream::empty; // core::posix::StandardStream::stdout | core::posix::StandardStream::stderr;
+          auto child = core::posix::fork([&]() {
+            auto grandchild = core::posix::exec(exe_path, args, env, flags);
+            grandchild.dont_kill_on_cleanup();
+            return core::posix::exit::Status::success;
+          }, flags);
+          child.wait_for(core::posix::wait::Flags::untraced);
+
+          DEBUG("Started session manager, will now try to connect ..");
+        }
+        catch (...) {
+          ERROR("Failed to start session manager instance");
+        }
+
+        std::this_thread::sleep_for(restart_interval);
+      }
     }
 
-    auto dispatcher = anbox::common::create_dispatcher_for_runtime(rt);
+    if (!stub) {
+      ERROR("Couldn't get a connection with the session manager");
+      return EXIT_FAILURE;
+    }
 
     bool success = false;
-
+    auto dispatcher = anbox::common::create_dispatcher_for_runtime(rt);
     dispatcher->dispatch([&]() {
       if (stub->ready()) {
-        try {
-          stub->launch(intent_, graphics::Rect::Invalid, stack_);
-          success = true;
-        } catch (std::exception &err) {
-          ERROR("err %s", err.what());
-        }
+        ss.reset();
+        success = try_launch_activity(stub);
         trap->stop();
         return;
       }
 
-      DEBUG("Android hasn't fully booted yet. Waiting a bit..");
+      DEBUG("Android hasn't fully booted yet. Waiting a bit ..");
 
       stub->ready().changed().connect([&](bool ready) {
         if (!ready)
           return;
-        try {
-          stub->launch(intent_, graphics::Rect::Invalid, stack_);
-          success = true;
-        } catch (std::exception &err) {
-          ERROR("Failed to launch activity: %s", err.what());
-          success = false;
-        }
+        ss.reset();
+        success = try_launch_activity(stub);
         trap->stop();
       });
     });
 
     boost::asio::deadline_timer timer(rt->service());
     timer.expires_from_now(max_wait_timeout);
     timer.async_wait([&](const boost::system::error_code&) {
-      WARNING("Stop waiting as we're already waiting for too long. Something is wrong");
-      WARNING("with your setup and the container may have failed to boot.");
+      WARNING("Stopped waiting as we're already waited for too long. Something");
+      WARNING("is wrong with your setup or the container has failed to boot.");
```

Comment:
Added a comment to explain where bugs can be filed. I will propose another PR soon which will add a few more debugging capabilities which I will link here as well.